#10_2 商品購入機能

Lubellum opened the pull request:
#What
商品購入機能の実装。

#Why
購入機能を実装する事で、商品の売買を可能とするため。

必要な動画及び画像は以下です。

* 必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/19f15e8f08aecd78a44e8a7ec8ddf0c6

* 入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/41f6d6f2948ce705f1515e09b388ecc4

* ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/9d752b1617ea1a8f86d89d70db360483

* ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/1a88be11465964f57fa84831f324f5c2


* ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/528b7e75ba52a463c656e620bb991970

* 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
https://gyazo.com/dbb1b23567a8a45aca7d49557ac7f52e

* 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/391c5fc5956027b3331256ecff0a4b9a

* ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/4a7d78b7aa2a83e9a34728f6348c2d54

* ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/e509744ac416458744a60660d14a45ee

* テスト結果の画像
https://gyazo.com/59e3f0c00b484b1967ed7c6a862999c4

Lubellum committed:
Formオブジェクトの作成

Lubellum committed:
バリデーションの記述

Lubellum committed:
Formオブジェクトの保存処理の記述

Lubellum committed:
purchase_recordsのアンダーバーを全角から半角へ修正

Lubellum committed:
保存先モデル名をpurchaseからpurchase_recordへ修正

Lubellum committed:
mailing_addresモデルの作成

Lubellum committed:
mailing_addresテーブルのマイグレーション実行

Lubellum committed:
デフォルトのビューファイル名ordersをmailing_addressesに変更及びルーティングの記述

Lubellum committed:
中間テーブルpurchase_recordsの作成

Lubellum committed:
各モデルのアソシエーションの記述

今回実装する全モデルが出揃ったため

Lubellum committed:
mailing_addreモデルを削除し、新たにmailing_addressモデルを作成

Lubellum committed:
外部キーの_idの部分を削除

外部キーを記述する場合は_idが不要。
外部キーであると判断され勝手に_idが付与される。

Lubellum committed:
ルーティングのスペルを修正

Lubellum committed:
mailing_addressモデルのアソシエーションを記述

Lubellum committed:
mailing_addressesコントローラーの作成

Lubellum committed:
ビューファイルのフォルダ名を「mailing_addreses」から「mailing_addresses」に修正

綴りが間違っており、上手くページ遷移が行われなかったため

Lubellum committed:
購入画面への遷移処理

Lubellum committed:
商品購入画面のフォームにシンボルを記述

Lubellum committed:
商品購入画面のパスの指定

Lubellum committed:
Formオブジェクトの保存の記述をnewからcreateに修正

newでは情報が保存されないため、createが正しい。

Lubellum committed:
Formオブジェクトの保存するデータ名を修正

user_idとitem_idは既に定義されているので、それらを元にした「購入情報」が含まれているpurchaseカラムを追記

Lubellum committed:
保存先をMilingAddressモデルからPurchaseMailingに修正

Lubellum committed:
saveメソッドのendを追記及びvalidatesの綴りを修正

Lubellum committed:
商品購入画面でエラーハンドリングの実施

Lubellum committed:
purchase_mailingクラスのテストコード準備

Lubellum committed:
FactoryBotの記述

Lubellum committed:
purchase_mailing_specの雛形作成

Lubellum committed:
Fakerで生成される電話番号の範囲を変更

Lubellum committed:
@purchase_mailingにユーザー情報と商品情報を紐付け

purchase_mailingクラスはモデルでは無いため、FactoryBot内ではなくbeforeアクションとして定義する必要がある

Lubellum committed:
電話番号のバリデーションを11桁のみから、10桁〜11桁を許可に変更

Lubellum committed:
テスト　建物名が無くても購入できる

Lubellum committed:
テスト　郵便番号が必須

Lubellum committed:
テスト　郵便番号にはハイフン(-)が必須

Lubellum committed:
テスト　郵便番号は半角文字の必要がある

Lubellum committed:
beforeアクション後、待機時間を追加

処理が高速すぎる事で 「Mysql2::Error: MySQL client is not connected」エラーが出てしまったため処理にラグを持たせて解決を図った。

Lubellum committed:
待機時間を0.01秒から0.1秒に変更

0.01秒では再度エラーが出てしまったため。

Lubellum committed:
テスト　郵便番号は「数字3桁-数字4桁」の必要がある(4桁-4桁)

Lubellum committed:
Fakerで作成される電話番号の範囲を変更

0_000_000_000では1文字目が0となった場合、電話番号が9桁にしかならずエラーが発生したため

Lubellum committed:
テスト　郵便番号は「数字3桁-数字4桁」の必要がある(3桁-5桁)

Lubellum committed:
テスト　都道府県の選択が必須

Lubellum committed:
都道府県選択のnumericalityを0から1に修正及びエラーメッセージの変更

Lubellum committed:
テスト　都道府県の選択はidが1以外の必要がある

Lubellum committed:
テスト　都道府県の選択は数字での入力が必須

Lubellum committed:
テスト　市区町村が必須

Lubellum committed:
テスト　番地が必須

Lubellum committed:
テスト　電話番号が必須

Lubellum committed:
テスト　電話番号は10桁〜11桁の必要がある(9桁)

Lubellum committed:
テスト　電話番号は10桁〜11桁の必要がある(12桁)

Lubellum committed:
テスト　電話番号は半角数字の必要がある(全角数字)

Lubellum committed:
テスト　電話番号は半角数字の必要がある(全角数字)

Lubellum committed:
テスト　userが紐付いていないと購入出来ない

Lubellum committed:
テスト　itemが紐付いていないと購入出来ない

Lubellum committed:
クレジットカードカラムのシンボル記述

Lubellum committed:
購入情報に商品idの紐付け

Lubellum committed:
決済機能に必要なファイルの作成

Lubellum committed:
外部キーのpurchase_record_idの記述を変更

カラム名がpurchase_record_idかつ外部キーなのでシンボルとして記述する際はpurchase_recordが正しい。

Lubellum committed:
商品購入情報に商品情報のmerge

mailing_addressesモデルはitemモデルにネストされ子モデルとなっているのでfind paramsの指定が通常とは異なるため

Lubellum committed:
t.timestampsの上に改行を挿入

可読性向上のため

Lubellum committed:
保存先カラム名をpurchase_record_idに修正

Lubellum committed:
tokenをpayjpに送信出来るようパラメーターに含める

Lubellum committed:
gem payjpのインストール

Lubellum committed:
商品購入画面の詳細表示

Lubellum committed:
tokenが必須のバリデーションを設定

Lubellum committed:
payjpの決済処理(JavaScript)

Lubellum committed:
payjpの決済処理(コントローラー)

Lubellum committed:
tokenについてのテストコードを記述

Lubellum committed:
webpackerにテスト公開鍵の設定を記述

Lubellum committed:
秘密鍵を環境変数に設定

Lubellum committed:
公開鍵を環境変数に設定

Lubellum committed:
Rubocopの実施

Lubellum committed:
インスタンス変数@itemを変数化

Lubellum committed:
payjpに送信する金額に関する記述を修正

Lubellum committed:
商品が売却済みかどうかでSOLD OUTを表示させるか分岐

Lubellum committed:
商品が出品しているかどうかで分岐処理

購入履歴がnilの場合は表示し、購入履歴がない場合は表示させない。

Lubellum committed:
商品が売却済みの場合、編集画面に遷移しようとするとリダイレクト

Lubellum committed:
Rubocopの実施

Lubellum committed:
商品が売却済みの場合、購入画面に遷移しようとするとリダイレクト

Lubellum committed:
ログアウト状態の場合、購入画面に遷移しようとするとリダイレクト

Lubellum committed:
詳細画面の値段の記述を修正

strongmasahiro commented:
レビュー依頼ありがとうございます。
```
ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
```
こちらの動画に関して、いただいた動画ですと**編集ページではなく購入ページに遷移しようとしている動画**になっております。

大変お手数ですが、再度以下の条件に合うように挙動確認の動画の提出をお願い致します。
```
ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
```

Lubellum commented:
大変申し訳ございません。

再度撮影した動画が以下となります。
* ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/1f5b7e84084577795fccbc11f8635533

yuhi-di commented on app/controllers/mailing_addresses_controller.rb:
```diff
@@ -0,0 +1,44 @@
+  before_action :authenticate_user!, only: [:index]
+  before_action :set_item, only: [:index, :create]
+
+  def index
```
以下の実装条件を満たすために修正をしましょう。
・自身が出品した商品の購入ページへ遷移しようとすると、トップページに遷移すること。

現状ですと、自信が出品した売却されていない商品の購入ページに遷移してしまいます。

yuhi-di commented on app/views/items/show.html.erb:
```diff
@@ -8,35 +8,35 @@
-      <%# 商品が売れていない場合はこちらを表示しましょう %>
-      <% if current_user.id != @item.user_id %>
-        <%= link_to "購入画面に進む", "#" ,class:"item-red-btn"%>
+        <%# 商品が売れていない場合はこちらを表示しましょう %>
+        <% if current_user.id != @item.user_id %>
+          <%= link_to "購入画面に進む", item_mailing_addresses_path(@item.id) ,class:"item-red-btn"%>
+        <% end %>
+        <%# //商品が売れていない場合はこちらを表示しましょう %>
```
34行目、38行目は実装案内に関するコメントになりますので、削除しておきましょう。
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめいたします。

yuhi-di reviewed (commented):
レビュー依頼ありがとうございます！
2点コメントさせていただきましたので、ご確認お願いいたします。

Lubellum committed:
自身が出品した商品の購入ページへ遷移しようとすると、トップページに遷移するよう修正

Lubellum committed:
ログインしているユーザーによって処理の分岐を行う箇所のコメントを修正。

一眼見た際、「if current_user.id == @item.user_id」なのか「if current_user.id != @item.user_id」分かりにくいため。

strongmasahiro commented on app/controllers/mailing_addresses_controller.rb:
```diff
@@ -0,0 +1,44 @@
+  before_action :set_item, only: [:index, :create]
+
+  def index
+    if current_user.id == @item.user_id
+      redirect_to root_path
+    else @item.purchase_record.nil?
+      @purchase_mailing = PurchaseMailing.new
+    end
```
```
    if current_user.id == @item.user_id
      redirect_to root_path
    else @item.purchase_record.nil?
      @purchase_mailing = PurchaseMailing.new
    end
```
こちらの記述で検証したところ、自身が出品していない売却済みの商品の購入ページに遷移できてしまいます。
以下を参考に修正しましょう！
```
def index
  if "出品者である" または "売却済みである"
   redirect_to root_path
  end
  @purchase_mailing = PurchaseMailing.new
end
```

strongmasahiro reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

Lubellum committed:
コメントの詳細を修正

コメント内容が適切では無かったため。

Lubellum committed:
自身が出品していない売却済みの商品の購入ページに遷移出来ないように修正

yuhi-di commented:
修正とレビュー依頼ありがとうございます。

しっかりと実装できていることが確認できましたので、
**LGTM**とさせていただきます！

こちらのブランチは忘れずにMergeを行っていただき、
ローカルのMasterにPullしておきましょう。

商品購入機能の実装、お疲れ様でした。
引き続き、応援しております。
